app/misc.py

```diff
@@ -688,29 +690,138 @@ def getInviteCodeInfo(uid):
 
     return info
 
+def start_forwarding_recent_unread_notifications(
+    period_sec=300, lower_bound_days=30
+):
+    """
+    Starts a 'daemon' greenlet that will run for
+    the duration of the program, but not longer,
+    while periodically checking for unread notifications
+    to forward by email to users who've opted-in to
+    this preference.
+    """
+
+    def watch_forward():
+        """
+        Periodically fetching and forwarding notifications
+        for user who've opted-in to the feature
+        """
+        get_post_link = lambda i: url_for("sub.view_post", sub=i.name, pid=i.pid)
+        get_comment_link = lambda i: url_for(
+                "sub.view_perm",
+                sub=i.sub,
+                pid=i.pid,
+                slug=slugify(i.title),
+                cid=i.cid,
+            )
+        
+        while True:
+            gevent.sleep(period_sec)
+            now = datetime.utcnow()
+            before = now - timedelta(days=lower_bound_days)
+            after = now - timedelta(seconds=period_sec)
+
+            notifs_cursor = (
+                Notification.select(
+                    Notification.type,
+                    Notification.sub,
+                    Notification.post,
+                    Notification.comment,
+                    Notification.target,
+                    Notification.created,
+                    Sub.name,
+                    SubPost.pid,
+                    SubPost.title,
+                    SubPost.link,
+                    SubPostComment.cid
+                )
+                .join(Sub)
+                .join(SubPost)
+                .join(SubPostComment)
+                .where(
+                    Notification.read.is_null(False) & before
+                    < Notification.created & Notification.created
+                    < after
+                )
+                .execute()
+            )
+
+            users_cursor = (
+                User.select(User.name, User.email)
+                .where(User.opt_in_email_forwarded_notifications == True)
+                .execute()
+            )
+
+            opt_ins = {user.uid: user for user in users_cursor}
```

Comment:
beware, this could be many many users at some point in the future. it's better to filter for this in the main query

Reply:
Right

Reply:
In fact I don't think I am able to get in a single query both:
- notifications; and
- user data for user who are the target of a notification. 
So I need to run two queries. But then if use the former as an argument for the latter, as in `where(..., User.uid << [notif.target for notif in notif_cursor])` that would consume the `notif_cursor` iterator. So I am not sure how to improve on this specific part.